README.md
# MultiConfig [![Build Status](https://secure.travis-ci.org/shadabahmed/multi_config.png)](https://secure.travis-ci.org/shadabahmed/multi_config)

## Description

This `acts_as` style extension provides the capabilities for using multiple database configuration files. Any model can
specify what database config file to use by using a Rails 3.2 style self.config_file= method.

## Installation

In your Gemfile:

    gem 'multi_config'

Or, from the command line:

    gem install multi_config

## Example

    class DifferentTable < ActiveRecord::Base
      self.config_file = 'other_db'
    end

    # In config/other_db.yml

    development: &development
      database: db/other_db
      host: localhost
      adapter: sqlite3

    test:
      <<: *development

## Notes
All database config files should be similar to the database.yml. You need to specify configuration for all the environments
in use like you would specify in database.yml.

ActiveRecord uses *ActiveRecord::Base.configurations* hash to store configurations. This extension modifies the *configurations*
hash by adding the configurations from the other database files. The keys for the configuration in the other config are namespaced.
For e.g. if you specify `other_db` config file with environments `test` and `development` then configurations hash would have
these configs in `other_db_test` and `other_db_development` keys.

If you need to create migrations for the other db, then you will have to make modifications in the migration file. See this
[guide](http://stackoverflow.com/questions/1404620/using-rails-migration-on-different-database-than-standard-production-or-devel) on
how to do that. The parameter for `establish_connection` would be the namespaced key for you config.

## Versions
All versions require Rails 3.0.x and higher. Tested for Rails 4

## Roadmap

1. Add support for generators where you can specify which db file to use in the rails generate command. Would mitigate the need to manually modify migrations.

## Contributing to `multi_config`

- Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
- Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
- Fork the project
- Start a feature/bugfix branch
- Commit and push until you are happy with your contribution
- Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
- Send me a pull request
- Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Development
 - Run tests - `rake`
 - Generate test coverage report - `rake coverage`. Coverage report path - coverage/index.html
 - Generate documentation - `rake rdoc`

## Copyright

Copyright (c) 2012 Shadab Ahmed, released under the MIT license


CHANGELOG.md
Unreleased ([changes](https://github.com/shadabahmed/multi_config/compare/v0.2.2...master))
-------------------
**Note: Not implemented yet**
  * Add generators for migrations

v0.2.2, 2013-07-07 ([changes](https://github.com/shadabahmed/multi_config/compare/v0.2.0...v0.2.2))
-------------------
  * Updated for rails 4. Only travis and gemfile needed to be changed

v0.2.0, 2012-10-23 ([changes](https://github.com/shadabahmed/multi_config/compare/v0.1.9...v0.2.0))
-------------------
  * Refactored specs

v0.1.9, 2012-09-18 ([changes](https://github.com/shadabahmed/multi_config/compare/v0.1.8...v0.1.9))
-------------------
  * Refactored class variables

v0.1.8, 2012-09-10 ([changes](https://github.com/shadabahmed/multi_config/compare/v0.1.6...v0.1.8))
-------------------
  * Test for initializers added. Coverage results now 100%
  * Refactoring of spec
  * More documentation
  * 100% rdoc documentation coverage

v0.1.6, 2012-09-09 ([changes](https://github.com/shadabahmed/multi_config/compare/v0.1.1...v0.1.6))
-------------------
  * Further refactoring
  * Tons of documentation
  * More test cases
  * Removed some methods into a module to prevent polluting ActiveRecord

v0.1.1, 2012-09-09 ([changes](https://github.com/shadabahmed/multi_config/compare/v0.1.0...v0.1.1))
-------------------
  * Build breaking because of a missing folder fixed

v0.1.0, 2012-09-09 ([changes](https://github.com/shadabahmed/multi_config/compare/v0.0.1...v0.1.0))
-------------------
  * Refactored quite a bit
  * 100% coverage for majority of the gem, except the railtie class.

v0.0.1, 2012-09-08 ([source](https://github.com/shadabahmed/multi_config/tree/53bf0b63f8e217379459b83b173c319039d6edc0))
-------------------
  * First commit with working code
  * Good test coverage


spec/spec_helper.rb
# Notice there is a .rspec file in the root folder. It defines rspec arguments

# Ruby 1.9 uses simplecov. The ENV['COVERAGE'] is set when rake coverage is run in ruby 1.9
if ENV['COVERAGE']
  require 'simplecov'
  SimpleCov.start do
    # Remove the spec folder from coverage. By default all code files are included. For more config options see
    # https://github.com/colszowka/simplecov
    add_filter File.expand_path('../../spec', __FILE__)
  end
end

# Modify load path so you can require 'multi_config' directly.
$LOAD_PATH.unshift(File.expand_path('../../lib', __FILE__))

require 'rubygems'
# Loads bundler setup tasks. Now if I run spec without installing gems then it would say gem not installed and
# do bundle install instead of ugly load error on require.
require 'bundler/setup'

# This will require me all the gems automatically for the groups. If I do only .setup then I will have to require gems
# manually. Note that you have still have to require some gems if they are part of bigger gem like ActiveRecord which is
# part of Rails. You can say :require => false in gemfile to always use explicit requiring
Bundler.require(:default, :test)
require 'active_record'

require 'multi_config/orms/active_record'
# This is required to rcov to run. Otherwise it runs without specs
require 'rspec/autorun'

# Requires supporting ruby files with custom matchers and macros, etc,
# in spec/support/ and its subdirectories.
Dir[File.join(File.dirname(__FILE__), "support/**/*.rb")].each { |f| require f }

# Set Rails environment as test
ENV['RAILS_ENV'] = 'test'

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# Require this file using `require "spec_helper"` to ensure that it is only
# loaded once.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = 'random'

  # Setting rails default settings such as Rails.root before every spec
  config.before(:each) do
    set_rails_conf
  end
end


spec/fixtures/models.rb
class UsesDatabaseYml < ActiveRecord::Base; end
class UsesOtherYml < ActiveRecord::Base; end

spec/multi_config_spec.rb
require 'spec_helper'
require 'rails/railtie'

describe MultiConfig::Railtie do
 describe "railtie initialization" do
    before do
      # Stub the initializer method and store the block to run later
      Rails::Railtie.stub(:initializer) { |initializer_name, &block| @railtie_initializer_block = block }
      # Stub  onload to yield directly
      ActiveSupport.stub(:on_load).and_yield
      # Remove the class definition
      MultiConfig.send(:remove_const, :Railtie)
    end
    it "inserts the module when railtie hook executes" do
      # Reload class definition so that class in instantiated and Railitie.initializer is called
      load 'multi_config.rb'
      MultiConfig::Railtie.should_receive(:insert).once()
      # Call the initializer block we stored earlier. We wait so that the file is completely loaded and method Railitie.insert
      # to be defined first. If we do not wait and yield directly from initializer, it would raise method not found exception
      # for Railtie.insert since we removed the constant
      @railtie_initializer_block.call
    end
  end

  describe '#insert' do
    it "calls included the module in ActiveRecord" do
      ActiveRecord::Base.should_receive(:send).with(:include, MultiConfig::ORMs::ActiveRecord).once()
      MultiConfig::Railtie.insert
    end
  end
end

lib/multi_config.rb
require "multi_config/version"

# Module for this gem
module MultiConfig
  # Checking if Rails::Railtie exists. Only then loca this railtie
  if defined? Rails::Railtie
    require 'rails'
    # Railtie class for the gem. Initialization magic happens here. Inherits from Rails::Railtie so that initializer
    # methods are available to it
    class Railtie < Rails::Railtie
      # Railtie initializer method
      initializer 'multi_config.active_record' do
        # When active_record is loaded, only then run this. See method documentation to learn mode
        ActiveSupport.on_load :active_record do
          # Notice that full namespace is not given since search for Railtie class begin from this Module itself.
          Railtie.insert
        end
      end
    end
  end

  class Railtie
    # Include the module MultiConfig::ORMs::ActiveRecord in ActiveRecord::Base class
    def self.insert
      # Even though ActiveSupport called this method only when ActiveRecord was loaded. We are just being extra safe.
      if defined?(ActiveRecord)
        require 'multi_config/orms/active_record'
        # Calling private method :include via send. This is typically used in extensions to include a module.
        ActiveRecord::Base.send(:include, MultiConfig::ORMs::ActiveRecord)
      end
    end
  end
end

lib/multi_config/orms/active_record.rb
module MultiConfig
  # This module encloses implementations for multiple ORMs
  module ORMs
    # Implementation for the ActiveRecord ORM
    module ActiveRecord
      # Method called when the module is included.
      # * +mod+ - Name of the class including this module.
      # Calls mod.extend with param ClassMethods so that methods in ClassMethods module become class methods of the class.
      def self.included(mod)
        mod.extend ClassMethods
        # Setting @@_multi_config_db_configs in the including class. This will help us keep track of database configuration files already included
        mod.send(:class_variable_set, :'@@_multi_config_db_configs', Hash.new { |h, k| h[k] = [] })
      end

      # This `acts_as` style extension provides the capabilities for using multiple database configuration files. Any
      # model can specify what database config file to use by using a Rails 3.2 style self.config_file= method.
      #
      # Example:
      #
      #   class DifferentTable < ActiveRecord::Base
      #     self.config_file = 'other_db'
      #   end
      #
      #   In config/other_db.yml
      #
      #   development: &development
      #     database: db/other_db
      #     host: localhost
      #     adapter: sqlite3
      #
      #   test:
      #     <<: *development
      module ClassMethods
        # Use the specified config file for database. You can specify file without .yml extension. If you specify database.yml or
        # database, it will not have any effect since that is loaded by default.
        def config_file=(file_name)
          file_name += '.yml' unless File.extname(file_name).eql? '.yml'
          # Load config file if it is not database.yml
          unless file_name == 'database.yml'
            namespace = File.basename(file_name, '.yml')
            # Update active_record configurations hash
            add_db_config(file_name, namespace)
            # Raise error if the config file does not have the current environment
            raise "Configuration for #{::Rails.env} environment not defined in #{Config.path file_name}" unless configurations.include? "#{namespace}_#{::Rails.env}"
            # Establish connection. This is the only way I found to different database config. Will try to find alternative
            establish_connection "#{namespace}_#{::Rails.env}"
          end
        end

        private
        # Adds namespaced database configuration to configurations hash.
        # * +file_name+ - file name of the database config file
        # * +namespace+ - this is the basename of the file. e.g. other_db when file is other_db.yml. Any other can be specified
        # If a db config file has been added to the configurations once with a specific namespace, it will not be loaded again for
        # same namespace. For .e.g. Models A and B specify config file 'other_db', then other_db.yml will be read only once.
        def add_db_config(file_name, namespace)
          # Read class var from the including class
          db_configs = class_variable_get(:'@@_multi_config_db_configs')

          # Don't do anything if the namespace has already been added
          unless db_configs.include?(namespace)
            configurations.merge!(Config.load(file_name, namespace))
            # Add class name in db_configs. This is to keep track of models that have specified a particular db config file.
            db_configs[namespace] << name
          end
        end
      end

      # Defines helper methods to be used the the extension. Since this is in the enclosing module for ClassMethods Module.
      # This will be searched first when Config is referenced. The Ruby Programming Language - Page 276
      module Config

        # Returns the absolute path for the file specified. The file is supposed to be located in the config folder.
        def self.path(file_name)
          File.join(Rails.root, 'config', file_name)
        end

        # Load a yaml file and modify the keys by appending <tt><namespace>_</tt> string in the keys. For e.g. if a yaml
        # file with namespace other_db is loaded with environments test and development, the returned hash from this method
        # would contain keys 'other_db_test' and 'other_db_development'.
        def self.load(file_name, namespace)
          begin
            require 'erb'
            YAML.load(ERB.new(IO.read(path(file_name))).result).inject({}) do |hash, (k, v)|
              hash["#{namespace}_#{k}"]=v
              hash
            end
          # ruby 1.9 raises SyntaxError exception which is not subclass of
          rescue StandardError, SyntaxError
            raise "File #{path file_name} does not exist in config" unless File.exists?(path(file_name))
            raise "Invalid config file #{path file_name}"
          end
        end
      end
    end
  end
end